packages/text-input/src/TextInput.tsx

```diff
@@ -0,0 +1,331 @@
+  z-index: 0;
+`;
+
+const inputStyle = css`
+  width: 400px;
+  height: 36px;
+  border-radius: 4px;
+  padding-left: 12px;
+  font-size: 14px;
+  font-weight: normal;
+  &:placeholder {
+    color: ${uiColors.gray.base};
+  }
+  &:focus {
+    outline: none;
+  }
+`;
+
+const errorIconStyle = css`
+  position: absolute;
+  right: 10px;
+  color: ${uiColors.red.base};
+`;
+
+const validIconStyle = css`
+  position: absolute;
+  right: 10px;
+  color: ${uiColors.green.base};
+`;
+
+const optionalStyle = css`
+  position: absolute;
+  right: 12px;
+  color: ${uiColors.gray.dark1};
+  font-size: 12px;
+  font-style: italic;
+  font-weight: normal;
+`;
+
+const errorMessageStyle = css`
+  color: ${uiColors.red.base};
+  font-size: 14px;
+  height: 20px;
+  padding-top: 4px;
+  font-weight: normal;
+`;
+
+/**
+ * # TextInput
+ *
+ * TextInput component
+ *
+ * ```
+<TextInput label='Input Label' onChange={() => execute when value of input field changes}/>
+```
+ * @param props.label Text shown in bold above the input element.
+ * @param props.description Text that gives more detail about the requirements for the input.
+ * @param props.optional Whether or not the field is optional.
+ * @param props.disabled Whether or not the field is currently disabled.
+ * @param props.onChange Callback to be executed when the value of the input field changes.
+ * @param props.placeholder The placeholder text shown in the input field before the user begins typing.
+ * @param props.errorMessage The message shown below the input field if the value is invalid.
+ * @param props.state The current state of the TextInput. This can be none, valid, or error.
+ * @param props.value The current value of the input field. If a value is passed to this prop, component will be controlled by consumer.
+ * @param props.setValue Callback used to set the value of the input field.
+ * @param props.className className supplied to the TextInput container.
+ */
+const TextInput = React.forwardRef(
+  (
+    {
+      label,
+      description,
+      optional = false,
+      disabled = false,
+      onChange,
+      placeholder,
+      errorMessage,
+      state,
+      value: controlledValue,
+      setValue: setControlledValue,
+      className,
+    }: TextInputProps,
+    forwardRef: React.Ref<HTMLInputElement>,
+  ) => {
+    const isControlled = typeof controlledValue === 'string';
+    const [uncontrolledValue, setUncontrolledValue] = useState('');
+    const value = isControlled ? controlledValue : uncontrolledValue;
+    const setValue = isControlled ? setControlledValue : setUncontrolledValue;
+    const { usingKeyboard: showFocus } = useUsingKeyboardContext();
+    const [hasFocus, setHasFocus] = useState(false);
+    const inputSelectorProp = createDataProp('input-selector');
+
+    function getInputColorFromState() {
+      if (state === State.None) {
+        return '#CCC';
+      }
+
+      return state === State.Valid ? uiColors.green.base : uiColors.red.base;
```

Comment:
I might suggest taking the concept of happy-path further in this function in a couple ways.

- First is to break out the ternary statement you're returning into separate return statements.
- Second is to make the final return statement the default state, and check for the valid and error states in the conditionals.

Example of what this would look like:
```js
function getInputColorFromState() {
  if (state === State.Error) {
    return uiColors.red.base;
  }

  if (state === State.Valid) {
    return uiColors.green.base;
  }

  return '#CCC';
}
```